### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by junderw**
This implements encode and decode.

tests include test vectors from the BOLT#11 and have checks for decoding, making sure the info is correct, then encoding and making sure the payment request is the same with both reconstruction AND signing (RFC6979 deterministic signing)

Help Wanted:

Anyone who can list of cases I should add for tests, as well as any refactoring help please feel free to commit directly to the `wip` branch.

Thanks.

### CODE REVIEW

**Comment by junderw** [comment]
Also, I copied convert from bech32, but there is currently a PR I have open to expose convert.

https://github.com/bitcoinjs/bech32/pull/18

Once this is merged I will replace convert with bech32.convert

File: payreq.js [hunk 1]
```diff
@@ -205,28 +205,22 @@ const purposeCommitEncoder = (data) => {
...
+  expire_time: 6, // default: 3600 (1 hour)
+  min_final_cltv_expiry: 24, // default: 9
+  fallback_address: 9,
+  routing_info: 3 // for extra routing info (private etc.)
 }
 
+const TAGNAMES = Object.keys(TAGCODES).reduce((final, currentKey) => {
+  final[TAGCODES[currentKey].toString()] = currentKey
+  return final
+}, {})
```
Comment (dcousens) [inline payreq.js]: Interesting trick.
`reduce` is less clear in this case though,  I think I'd go with a `forEach` and `cosnt TAGNAMES = {}`

Comment (fanatid) [reply payreq.js]: `for-in` & `Object.keys` is different things

```js
const TAGNAMES = Object.keys(TAGCODES).reduce((final, currentKey) => {
  return Object.assign(final, { [TAGCODES[currentKey].toString()]: currentKev })
}, {})
```
3-line solution 😛

Comment (dabura667) [reply payreq.js]: https://stackoverflow.com/questions/25052758/for-in-vs-object-key-foreach-without-inherited-properties

I think I’ll do a more traditional way that’s more readable.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,118 @@
...
+  * coinType should be the name of a network object from bitcoinjs-lib
+* Alternatively: You can pass the result of decode into encode and it will use the
+signature and recoveryFlag attributes to reconstruct the payment request. In this
+case you will require also `coinType` and `timestamp` as well as all tags in the
+exact order of the original signed request.
+  * It is also required to pass the `payeeNodeKey` attribute when encoding an
+  already signed request, as decoders will recover the pubkey, any incorrect data
+  would cause an incorrect pubkey to be generated and will cause an error on the
+  decoding end when trying to send.
+* Note: `satoshis` can be defined up to 3 decimal points (ie. `0.001` is the smallest valid
```
Comment (dcousens) [inline README.md]: Old?

**Comment by dcousens** [comment]
Travis has been enabled on the repo,  but unclear if that PR needed to merge master to work here.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,138 @@
...
+  ],
+  "timestamp": 1496314658,
+  "timestampString": "2017-06-01T10:57:38.000Z",
+  "wordsTemp": "temp1pvjluezhp58yjmdan79s6qqdhdzgynm4zwqd5d7xmw5fk98klysy043l2ahrqspp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqfppqw508d6qejxtdg4y5r3zarvary0c5xw7kepvrhrm9s57hejg0p662ur5j5cr03890fa7k2pypgttmh4897d3raaq85a293e9jpuqwl0rnfuwzam7yr8e690nd2ypcq9hlkdwdvycqawny4p"
+}
+*/
+```
+
+### Encoding
+* MINIMUM NEED: `privateKey` and one `payment_hash` tag as well as one `description`
```
Comment (alexbosworth) [inline README.md]: > MINIMUM NEED: `privateKey` and one `payment_hash` tag as well as one `description`

> `description` defaults to empty string

Confusing, these seem to be at odds with each other

Comment (dabura667) [reply README.md]: Good point.

Comment (junderw) [reply README.md]: True.

What I mean by it is "the spec requires a signature, a payment hash and a description. The only one I can make an assumption for is description which we can default to an empty string."

Not sure if I should throw an error if description is missing... blegh...

Comment (dcousens) [reply README.md]: @junderw if the spec supports an empty string... support the default `''`

Comment (junderw) [reply README.md]: @dcousens ok done! (that's how it already was)

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+sudo: false
+language: node_js
+node_js:
+  - "4"
+  - "5"
+  - "6"
+  - "7"
+  - "8"
```
Comment (dcousens) [inline .travis.yml]: needs 8+

Comment (junderw) [reply .travis.yml]: fixed :-D

**Comment by dcousens** [comment]
Still neeed to review this in depth,  but,  no use holding back development in this PR.

**Pull request merged**